src/iotjs_string.c

```diff
@@ -98,57 +78,45 @@ bool iotjs_string_is_empty(const iotjs_string_t* str) {
 void iotjs_string_make_empty(iotjs_string_t* str) {
   IOTJS_VALIDATED_STRUCT_METHOD(iotjs_string_t, str);
 
-  if (_this->data != iotjs_empty_string) {
+  if (_this->data != NULL) {
     iotjs_buffer_release(_this->data);
-    _this->cap = _this->size = 0;
-    _this->data = iotjs_empty_string;
+    _this->size = 0;
+    _this->data = NULL;
   }
 }
 
 
 void iotjs_string_append(iotjs_string_t* str, const char* data, int size) {
   IOTJS_VALIDATED_STRUCT_METHOD(iotjs_string_t, str);
 
-  const int kCapacityIncreaseFactor = 2;
-
-  IOTJS_ASSERT(_this->data != NULL);
-  IOTJS_ASSERT(_this->cap >= _this->size);
+  IOTJS_ASSERT(data != NULL);
 
   if (size < 0) {
     size = strlen(data);
   }
+
   if (size == 0) {
     return;
   }
 
-  if (_this->cap == 0) {
-    // No buffer was allocated.
-    IOTJS_ASSERT(_this->size == 0);
-    IOTJS_ASSERT(_this->data == iotjs_empty_string);
-    _this->cap = _this->size = size;
-    _this->data = iotjs_buffer_allocate(_this->cap + 1);
-    strncpy(_this->data, data, _this->size);
-  } else if (_this->cap >= _this->size + size) {
-    // Have enough capacity to append data.
-    strncpy(_this->data + _this->size, data, size);
-    _this->size += size;
+  if (_this->data != NULL) {
+    _this->data = iotjs_buffer_reallocate(_this->data, _this->size + size);
```

Comment:
Are you aware that memory allocation can be as slow as garbadge colector? This code will incrase number of allocator calls. Let's assume that you string with 100 letters will be constructed from 10 words with 9 letters and space. Your code uses 10 allocations but old code used to call malloc 5 times. It is very significant difference.

Please consider keeping capacity. Why don't we store size and capacity in 2 x 16 bits. We will support strings with maximum size 65535 but do we need longer strings in IoT solution? 

I created PR for this change here. Please review it:
https://github.com/Samsung/iotjs/pull/638